Documentation/anti-patterns.md

```diff
@@ -0,0 +1,180 @@
+Similarly, if a client uses `WithReturnConnectionError` to return a connection
+error if dialing fails, it may miss opportunities to recover from transient
+network issues that are resolved shortly after the initial dial attempt.
+
+## Best practices for error handling in gRPC
+
+Instead of relying on failures at dial time, we strongly encourage developers to
+rely on errors from RPCs. When a client makes an RPC, it can receive an error
+response from the server. These errors can provide valuable information about
+what went wrong, including information about network issues, server-side errors,
+and incorrect usage of the gRPC API.
+
+By handling errors from RPCs correctly, developers can write more reliable and
+robust gRPC applications. Here are some best practices for error handling in
+gRPC:
+
+- Always check for error responses from RPCs and handle them appropriately.
+- Use the `status` field of the error response to determine the type of error that
+  occurred.
+- When retrying failed RPCs, use a backoff strategy to avoid
+  overwhelming the server or exacerbating network issues.
+- Avoid using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`,
+  as these options can introduce race conditions and result in unreliable and
+  difficult-to-debug code.
+
+### Example: Handling errors from an RPC
+
+The following code snippet demonstrates how to handle errors from an RPC in
+gRPC:
+
+```go 
+ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+defer cancel()
+
+res, err := client.MyRPC(ctx, &MyRequest{})
+if err != nil {
+    // Handle the error appropriately,
+    // log it & return an error to the caller, etc.
+    log.Printf("Error calling MyRPC: %v", err)
+    return nil, err
+}
+
+// Use the response as appropriate 
+log.Printf("MyRPC response: %v", res)
+
+```
+
+To determine the type of error that occurred, you can use the status field of
+the error response:
+
+
+```go 
+resp, err := client.MakeRPC(context.Background(), request) 
+if err != nil {
+  status, ok := status.FromError(err) 
+  if ok {
+    // Handle the error based on its status code 
+    if status.Code() == codes.NotFound {
+      log.Println("Requested resource not found")
+    } else {
+      log.Printf("RPC error: %v", status.Message())
+    }
+  } else {
+    //Handle non-RPC errors 
+    log.Printf("Non-RPC error: %v", err)
+  }
+  return
+}        
+
+// Use the response as needed 
+log.Printf("Response received: %v", resp) 
+```
+
+### Example: Using a backoff strategy
+
+
+When retrying failed RPCs, use a backoff strategy to avoid overwhelming the
+server or exacerbating network issues:
+
+
+```go 
+var res *MyResponse
+var err error
+
+// Retry the RPC call a maximum number of times
+for i := 0; i < maxRetries; i++ {
+    // Create a context with a timeout of one second
+    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+    defer cancel()
```

Comment:
The user should already have a context with a deadline.  In case they don't, this should probably happen _before_ the for loop to put bounds on the entire operation, not individual calls.